src/background/utils/icon.js

```diff
@@ -111,13 +117,36 @@ function updateState(tab, url = tab.url) {
   }
 }
 
-function setIcon(tab = {}, data = {}) {
+async function setIcon(tab = {}, data = {}) {
   // modern Chrome and Firefox use 16/32, other browsers may still use 19/38 (e.g. Vivaldi)
   const mod = data.blocked && 'b' || !isApplied && 'w' || '';
-  browserAction.setIcon({
-    path: Object.assign({}, ...[16, 19, 32, 38].map(n => ({
-      [n]: `/public/images/icon${n}${mod}.png`,
-    }))),
+  const iconData = {};
+  const iconOptions = {
     tabId: tab.id,
+    [iconCache ? 'imageData' : 'path']: iconData,
+  };
+  for (const n of [16, 19, 32, 38]) {
+    const path = `/public/images/icon${n}${mod}.png`;
+    iconData[n] = !iconCache ? path : iconCache[path] || await loadImageData(path);
+  }
+  browserAction.setIcon(iconOptions);
+}
+
+function loadImageData(path) {
```

Comment:
We can use `memoize` from `#/common` instead of implementing cache repeatedly.

Reply:
So far memoize seems to be used only once so maybe it's better to remove it there :-)
I'm not convinced we need an abstraction for something so simple.

Reply:
I think it's not about simple or not. By implementing locally, you need to define a variable outside the function and read / write cache in the original function, which is kind of dirty.

For me cache should be something that can easily opt in or out, without affecting the original function. So `loadImageData` should work exactly as `cachedLoadImageData`, and the user who uses this function need not to know about it.

Reply:
> So far memoize seems to be used only once

Maybe my old code is badly designed, or there was little chance to cache a function permanently like this.

Reply:
Okay, I'll use memoize then.

Reply:
Uhm, I don't like this.
`memoize` stores a resolved Promise in its cache so each time `await` will run internally.
And I'll still have to declare a separate variable like isChrome or isIconCacheEnabled.

Reply:
I'll just move updating of the cache inside setIcon.